File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    // int adj(vector<int>& nums,int n,vector<int> &dp){
    //     if(n==0) return nums[0];
    //     if(n<1) return 0;
    //     if(dp[n]!=-1) return dp[n];
    //     int pick=nums[n]+adj(nums,n-2,dp);
    //     int notpick=0+adj(nums,n-1,dp);
    //     return dp[n]=max(pick,notpick);
    // }
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size()+1,-1);
        // return adj(nums,nums.size()-1,dp);
        // dp[0]=nums[0];
        int prev=nums[0];
        int prev2=0;
        for(int i=1;i<nums.size();i++){
            int pick=nums[i];
            if(i>1) pick=nums[i]+prev2;
            int notpick=prev+0;
            int curi=max(pick,notpick);
            prev2=prev;
            prev=curi;
        }
        return prev;
    }
};

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp,-1);
        return fun(nums,nums.length-1,dp);
    }
    public int fun(int nums[],int ind,int[] dp){
        if(ind<0) return 0;
        
        if(dp[ind]!=-1) return dp[ind];
        int take = 0,notake = 0;
        take = nums[ind] + fun(nums,ind-2,dp);
        notake = 0 + fun(nums,ind-1,dp);
        
        return dp[ind] = Math.max(take,notake);
    }
}